### PR TITLE
Csharp Razor includes ZeroWidth NextToken

### PR DESCRIPTION
**Pull request by basoundr**
Fixes an internal bug 1190278. Chsarp Razor when looking for the next token it should also include zero width token so it can find EOFToken. Test will be added in the closed Test source

### CODE REVIEW

**Comment by basoundr** [comment]
Reviewers: @Pilchie @jasonmalinowski @dpoeschl @rchande @balajikris @brettfo @jmarolf @davkean

File: src/Workspaces/Core/Portable/Formatting/Rules/BaseIndentationFormattingRule.cs [hunk 1]
```diff
@@ -187,7 +187,7 @@ private static TextSpan GetSpanFromTokens(TextSpan span, SyntaxToken token1, Syn
             // current token is inside of the given span, get next token's start position.
             if (token2.Span.End <= span.End)
             {
-                token2 = token2.GetNextToken();
+                token2 = token2.GetNextToken(includeZeroWidth: true);
```
Comment (dpoeschl) [inline src/Workspaces/Core/Portable/Formatting/Rules/BaseIndentationFormattingRule.cs]: Why is this needed on GetNextToken() here, but not on GetPreviousToken() above? If the difference is intentional, it's quite subtle -- please add some comments.

**Comment by dpoeschl** [comment]
How does this impact other scenarios (VB or non-Razor) scenarios? Also, lots of features call GetNextToken with includeZeroWidth = false, do they need to be updated for this Razor scenario as well (in other words, will they still crash even though formatting is fixed now)?

**Comment by dpoeschl** [comment]
Can you please include an example of what the C# Razor code looks like that causes this problem. I know it will be included in the closed-source tests, but it's hard to reason about this change here without an example.

**Comment by jasonmalinowski** [comment]
And to @dpoeschl's good question: can an open-side test still exist with the problematic code?

**Comment by dpoeschl** [comment]
Tagging @heejaechang in case he's interested.

**Comment by rchande** [comment]
I thought @dpoeschl wrote some code that would synthesize from markup the Razor buffergraph.

**Comment by dpoeschl** [comment]
@rchande - There's [CreateProjectionBufferDocument](http://source.roslyn.io/#Roslyn.Services.Editor.UnitTests/Workspaces/TestWorkspace.cs,b86e6cfcf33ff79b). If it's just a matter of creating an appropriate projection and formatting, then we should be able to unit test that. However, it may be tricky if the specific Razor/Roslyn formatting handshake is involved here.

File: src/Workspaces/Core/Portable/Formatting/Rules/BaseIndentationFormattingRule.cs (hunk 1 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Formatting/Rules/BaseIndentationFormattingRule.cs]: are we missing formatting trivia attached to end of file token?

File: src/Workspaces/Core/Portable/Formatting/Rules/BaseIndentationFormattingRule.cs [hunk 2]
```diff
@@ -187,7 +187,7 @@ private static TextSpan GetSpanFromTokens(TextSpan span, SyntaxToken token1, Syn
...
                 end = token2.Span.Start;
                 if (token2.RawKind == 0)
```
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Formatting/Rules/BaseIndentationFormattingRule.cs]: this line is strange. and also line above.

I think it should be

var previous/nextToken = token1.GetPreviousToken()/GetNextToken();
...
if (preview/nextToken.RawKind == 0)
{
start/end = token1/2.FullSpan.Start/End;
}

Comment (basoundr) [inline src/Workspaces/Core/Portable/Formatting/Rules/BaseIndentationFormattingRule.cs]: @heejaechang I am not entirely sure if this would work because there might be cases where we might need to format the trivia which part of the visible area. Although, I dont have a repro  which explains the above mentioned case

File: src/Workspaces/Core/Portable/Formatting/Rules/BaseIndentationFormattingRule.cs (hunk 1 above)
Comment (basoundr) [inline src/Workspaces/Core/Portable/Formatting/Rules/BaseIndentationFormattingRule.cs]: @dpoeschl  @heejaechang We dont do it for GetPreviousToken() because the auto generated C# buffer content always have a well formed C# syntax in the beginning of the file. Theoretically speaking, there can be a crash through the GetPreviousToken calculation as well but practically it is impossible(No repro case).

**Comment by basoundr** [comment]
@dpoeschl @jasonmalinowski @rchande  [CreateProjectionBufferDocument](http://source.roslyn.io/#Roslyn.Services.Editor.UnitTests/Workspaces/TestWorkspace.cs,b86e6cfcf33ff79b) is it. Will update with the testcase on the open side.

File: src/EditorFeatures/CSharpTest/Formatting/Indentation/SmartIndenterTests.cs [hunk 3]
```diff
@@ -2149,6 +2149,81 @@ static void Main(string[] args)
                 expectedIndentation: BaseIndentationOfNugget + 4);
         }
 
+        [WorkItem(1190278)]
+        [WorkItem(3625, "https://github.com/dotnet/roslyn/pull/3625")]
```
Comment (dpoeschl) [inline src/EditorFeatures/CSharpTest/Formatting/Indentation/SmartIndenterTests.cs]: I don't think we normally link to the PR.

**Comment by basoundr** [comment]
@dpoeschl  I looked into the other calls for GetNextToken() with includeZeroWidth set to 'false' and it seems those are harmless. This particular bug affects both cshtml and vbhtml files. I will update the tests with a VB example as well.

**Comment by basoundr** [comment]
Summarizing the replies for the doubts asked above:

1. GetPreviousToken() is not an issue because the auto generated C# buffer content always have a well formed C# syntax in the beginning of the file.
2. I checked the other callsites of GetNexToken with includeZeroWidth set to false. They seem harmless
3. I played around with the projection buffer set to the crashing repro in a VS instance with the fix. There are no crashes.
4. Since we have testcases for both C# and VB in the open, we dont need an integration test in the closed source.

Reviewers: @Pilchie @jasonmalinowski @dpoeschl @rchande @balajikris @brettfo @jmarolf @davkean

**Comment by basoundr** [comment]
Reviewers: @Pilchie @jasonmalinowski @dpoeschl @rchande @balajikris @brettfo @jmarolf @davkean

**Comment by heejaechang** [comment]
:+1:
